lib/tests/testscripts/base_testscript.rb

```diff
@@ -347,33 +359,48 @@ def call_assertion(method, warned, *params)
 
       def replace_variables(input)
         return nil if input.nil?
+        return input unless input.include?('${')
 
         @testscript.variable.each do |var|
-          if !var.headerField.nil?
-            variable_source_response = @response_map[var.sourceId]
-            variable_value = variable_source_response.response[:headers][var.headerField]
-            input.gsub!("${" + var.name + "}", variable_value)
-          elsif !var.path.nil?
-
-            resource_xml = nil
-            
-            variable_source_response = @response_map[var.sourceId]
-            unless variable_source_response.nil?
-              resource_xml = variable_source_response.try(:resource).try(:to_xml) || variable_source_response.body
-            else
-              resource_xml = @fixtures[var.sourceId].to_xml
+          if input.include? "${#{var.name}}"
+            variable_value = nil
+
+            if !var.headerField.nil?
+              variable_source_response = @response_map[var.sourceId]
+              headers = variable_source_response.response[:headers]
+              headers.each do |key,value|
+                variable_value = value if key.downcase == var.headerField.downcase
+              end
+            elsif !var.path.nil?
+              resource_xml = nil
+
+              variable_source_response = @response_map[var.sourceId]
+              unless variable_source_response.nil?
+                resource_xml = variable_source_response.try(:resource).try(:to_xml) || variable_source_response.body
+              else
+                resource_xml = @fixtures[var.sourceId].to_xml
+              end
+
+              variable_value = extract_xpath_value(resource_xml, var.path)
             end
 
-            extracted_value = extract_xpath_value(resource_xml, var.path)
-
-            input.gsub!("${" + var.name + "}", extracted_value) unless extracted_value.nil?
+            unless variable_value
+              if var.defaultValue
+                variable_value = var.defaultValue
+              else
+                variable_value = ''
+              end
+            end
 
-          end if input.include? "${" + var.name + "}"
+            input.gsub!("${#{var.name}}", variable_value)
+          end
         end
 
         if input.include? '${'
-          puts "An unknown variable was unable to be replaced: #{input}!"
-          warning {  assert !input.include?('${'), "An unknown variable was unable to be substituted: #{input}" }
+          unknown_variables = input.scan(/(\$\{)([A-Za-z0-9]+)(\})/).map{|x|x[1]}
```

Comment:
Are alphanumeric values the only legal values in a variable? e.g. no underscores, etc.? I was looking at the `TestScript` details on the FHIR site, but couldn't find any concrete details one way or the other.